design-proposals/klip-2-produce-data.md

```diff
@@ -0,0 +1,347 @@
+  SQL standard and adds more complexity for the user to understand.
+
+### Code Enhancements
+
+KSQL in interactive mode is already configured to produce data to a command topic. This KLIP can
+leverage that existing functionality, which will require piping through various configurations into
+the `ServiceContext`. Beyond that, we must introduce a new command and Statement type to reflect the
+new functionality. 
+
+In interactive mode, this will be handled without distributing the statement to other servers. This
+operation will not be available in headless or embedded mode for the first version of the feature.
+
+### Schema Validation
+
+`INSERT INTO ... VALUES` will only support inserting into values for existing KSQL Streams/Tables,
+which already have a schema associated with them. Before producing a record to the underlying topic,
+the values for the fields will be verified against the schema. 
+
+### Serialization Support
+
+JSON, AVRO and DELIMITED value formats will all be supported, the latter of which will not accept
+complex types. Since we know the schema of the inserts, we can initially load the values into a 
+connect data type. We already have the proper serializers to convert from connect to each of the
+supported serialization formats.
+
+If the user specifies `AVRO` encoding in their `CREATE` statement, we will register the topic with 
+the schema registry proactively if the topic did not previously exist. Note that this will not 
+change how we handle `CREATE` statements for topics that already exist (today we do not validate
+that the schema supplied matches the schema in Schema Registry).
+
+### ROWKEY and ROWTIME Handling
+
+The underlying data and the way it is represented in KSQL sometimes differ, and this `INSERT VALUES`
+functionality needs to bridge that gap. For example take the following KSQL statement:
+```sql
+CREATE TABLE bar (id VARCHAR, foo INT) WITH (key="id")
+```
+
+In this case, we expect kafka records that resemble 
+```json
+{"key": "bar", "timestamp": 1, "value": {"id": "bar", "foo": 123}}
+```
+
+If this is not the case (e.g. the value of `"key"` and of `"value"->"id"` do not match), then the
+KSQL behavior is not well defined. 
+
+The column schema will always include the implicit `ROWKEY` and `ROWTIME` fields, and the user can
+choose to specify or ignore these fields. To prevent the inconsistencies described above, though,
+we will automatically enforce consistency whenever the key property is set in the `WITH` clause.
+
+* If the user sets a `KEY` property, they do not need to supply `ROWKEY` in their `INSERT INTO`
+statement (e.g. `INSERT INTO foo (id, foo) VALUES ("key", 123);`). In this case, the kafka message
+produced will automatically populate `ROWKEY` to the value of `id`: 
+  ```json
+  {"key": "key", "timestamp": 1, "value": {"id": "key", "foo": 123}}
+  ```
+* Analogous to the above, if the user provides the key but not `ROWKEY`, it will be populated for 
```

Comment:
nit: the above statement already provides the id, but not `ROWKEY`.  The analogous example would be the user providing ROWKEY and not ID... though I'm not sure that's possible...

Reply:
In the above statement, `ROWKEY` is the `key` field in the json

Reply:
Ernm...  yeah, I realize that.

So what I'm saying is that the statement in the example:

```sql
INSERT INTO foo (id, foo) VALUES ("key", 123);
```
Provides a value for 'id' and no value for 'ROWKEY'. (So KSQL puts 'key' into ROWKEY as well).

But then you state:

> Analogous to the above, if the user provides the key but not `ROWKEY`

Which I'm assuming is saying 

    if the user executes an INSERT statement and proves the field marked as the 'KEY' in the WITH clause, but not the 'ROWKEY',...

Which is exactly what the example statement _is_ doing.  Not sure which bit I'm misinterpreting!  It's not a biggie, but if it's confusing me it may be confusing others.

Reply:
Oh man 🤦‍♂️ I just misread what I wrote. Yes I see what you are saying...

Reply:
Cool.